### PR TITLE
Create abstraction layer for dart:io's Process commands

### PR DESCRIPTION
**Pull request by tvolkert**
With this change, they're run via instance methods on an object
obtained through the context. This will allow us to substitute
that object in tests with replay/record versions to allow us to
mock out the os-layer in tests.

### CODE REVIEW

**Comment by tvolkert** [comment]
@johnmccutchan

File: packages/flutter_tools/lib/executable.dart [hunk 1]
```diff
@@ -108,6 +109,7 @@ Future<Null> main(List<String> args) async {
     context.putIfAbsent(IOSSimulatorUtils, () => new IOSSimulatorUtils());
     context.putIfAbsent(SimControl, () => new SimControl());
     context.putIfAbsent(Usage, () => new Usage());
+    context.putIfAbsent(ProcessManager, () => new ProcessManager());
```
Comment (johnmccutchan) [inline packages/flutter_tools/lib/executable.dart]: You need to add the same putIfAbsent into test/src/context.dart

Comment (tvolkert) [inline packages/flutter_tools/lib/executable.dart]: Done

**Comment by johnmccutchan** [comment]
LGTM after figuring out travis failures

**Comment by eseidelGoogle** [comment]
@danrubel

File: packages/flutter_tools/lib/src/base/process_manager.dart [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+      executable,
+      arguments,
+      workingDirectory: workingDirectory,
+      environment: environment,
+      stdoutEncoding: stdoutEncoding,
+      stderrEncoding: stderrEncoding,
+    );
+  }
+
+  ProcessResult runSync(
```
Comment (danrubel) [inline packages/flutter_tools/lib/src/base/process_manager.dart]: I would prefer if we only used the `run` method throughout flutter tools and deleted this `runSync` method.

**Review by danrubel — APPROVED** [review]
LGTM with suggestion for you to address as you see fit.

File: packages/flutter_tools/lib/src/base/process_manager.dart (hunk 2 above)
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/process_manager.dart]: I like it, but it'd be a pretty invasive change, so let's do it in a different change.

**Comment by tvolkert** [comment]
I had to change around how context overrides are applied in the tests in order to fix the Travis failures.  PTAL.

File: packages/flutter_tools/lib/src/base/process_manager.dart (hunk 2 above)
Comment (tvolkert) [inline packages/flutter_tools/lib/src/base/process_manager.dart]: Filed https://github.com/flutter/flutter/issues/7102

**Comment by johnmccutchan** [comment]
I like the new test overrides interface. LGTM.

**Pull request merged**